# РК2
Фень Н.Т. ИУ5Ц-82Б
Вариант 31

In [7]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, f1_score

In [9]:
# Загрузка данных
df = pd.read_csv("Default_Fin.csv")

In [10]:
# Удаление неинформативного столбца
df = df.drop(columns=['Index'])

In [11]:
# Удаление строк с пропущенными значениями
df = df.dropna()

In [12]:
# Отделение признаков и целевой переменной
X = df.drop(columns=['Defaulted?'])
y = df['Defaulted?']

In [13]:
# Масштабирование признаков
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

In [14]:
# Разделение на обучающую и тестовую выборки
X_train, X_test, y_train, y_test = train_test_split(
    X_scaled, y, test_size=0.2, random_state=42
)

In [15]:
# Обучение модели SVM
svm_model = SVC(kernel='rbf', probability=True, random_state=42)
svm_model.fit(X_train, y_train)
svm_pred = svm_model.predict(X_test)

In [16]:
# Обучение модели Random Forest
rf_model = RandomForestClassifier(n_estimators=100, random_state=42)
rf_model.fit(X_train, y_train)
rf_pred = rf_model.predict(X_test)

In [17]:
# Оценка качества моделей
svm_accuracy = accuracy_score(y_test, svm_pred)
rf_accuracy = accuracy_score(y_test, rf_pred)
svm_f1 = f1_score(y_test, svm_pred)
rf_f1 = f1_score(y_test, rf_pred)

In [18]:
# Вывод результатов
results = pd.DataFrame({
    'Model': ['SVM', 'Random Forest'],
    'Accuracy': [svm_accuracy, rf_accuracy],
    'F1 Score': [svm_f1, rf_f1]
})

In [19]:
print(results)

           Model  Accuracy  F1 Score
0            SVM    0.9680  0.255814
1  Random Forest    0.9645  0.360360


### Используемые метрики

- **Accuracy (точность)** — показывает долю правильно предсказанных наблюдений. Подходит при относительно сбалансированных классах.
- **F1 Score** — гармоническое среднее между precision и recall. Особенно полезна при несбалансированных классах, так как учитывает как ложные срабатывания, так и пропущенные случаи.

### Вывод

- Обе модели показали высокую точность (>96%), что говорит о хорошем общем качестве классификации.
- Однако Random Forest имеет значительно более высокий F1 Score, что говорит о лучшем балансе между precision и recall.
- Следовательно, **модель Random Forest предпочтительнее**
